src/main/java/pl/edu/agh/firecell/engine/algorithm/SmokePropagator.java

```diff
@@ -0,0 +1,129 @@
+        if (westCell.isPresent()) smokeFromFire += westCell.get().material().smokeCoe() / 4;
+        if (eastCell.isPresent()) smokeFromFire += eastCell.get().material().smokeCoe() / 4;
+
+        double smokeDifference = getSmokeIndicatorDifference(oldCell, oldState, cellIndex);
+        if(Math.min(oldCell.smokeIndicator() + smokeDifference + smokeFromFire, 100) < 0.0005)
+            return 0.0;
+        return Math.min(oldCell.smokeIndicator() + smokeDifference + smokeFromFire, 100);
+    }
+
+    private double getSmokeIndicatorDifference(Cell oldCell, State oldState, Vector3i cellIndex) {
+        Int counter = new Int(oldCell.smokeIndicator());
+        return NeighbourUtils.neighboursStream(cellIndex)
+                .filter(oldState::hasCell)
+                .filter(neighbourIndex -> oldState.getCell(neighbourIndex).isFluid())
+                .map(neighbourIndex -> {
+                    double diff = 0.0;
+                    if (NeighbourUtils.up(cellIndex).equals(neighbourIndex)) {
+                        diff = -Math.min(oldCell.smokeIndicator(), 100 - oldState.getCell(neighbourIndex).smokeIndicator());
+                        counter.v += diff;
+                    } else if (NeighbourUtils.down(cellIndex).equals(neighbourIndex)) {
+                        diff = Math.min(oldState.getCell(neighbourIndex).smokeIndicator(),
+                                100 - oldCell.smokeIndicator());
+                        counter.v += diff;
+                    } else if (isHorizontal(cellIndex, neighbourIndex)) {
+                        int numberOfNeighboursMine = N(oldState, cellIndex).size() + 1;
+                        int numberOfNeighboursHim = N(oldState, neighbourIndex).size() + 1;
+                        if (cellAboveCanNotTakeSmoke(oldState, cellIndex)) {
+                            diff -= Math.min((100 - oldState.getCell(neighbourIndex).smokeIndicator()) / numberOfNeighboursHim,
+                                    (counter.v) / numberOfNeighboursMine);
+                        }
+                        if (cellAboveCanNotTakeSmoke(oldState, neighbourIndex)) {
+                            diff += Math.min((100 - counter.v) / numberOfNeighboursMine,
+                                    (oldState.getCell(neighbourIndex).smokeIndicator()) / numberOfNeighboursHim);
+                        }
+                    }
+                    return diff;
+                }).mapToDouble(Double::doubleValue).sum();
+    }
+
+    private boolean cellAboveCanNotTakeSmoke(State state, Vector3i index) {
+        Vector3i cellAbove = NeighbourUtils.up(index);
+        return (state.hasCell(cellAbove) && state.getCell(cellAbove).smokeIndicator() >= 50) ||
+                (state.hasCell(cellAbove) && !state.getCell(cellAbove).isFluid()) ||
+                !state.hasCell(cellAbove);
+    }
+
+    private List<Cell> N(State state, Vector3i cellIndex) {
```

Comment:
'N' is horrible name. Make it 'horizontalNeighboursWithSmokeCapacity()`

Reply:
Also, you use this method and always take the `.size()` of result list. Make it return this size then, and rename for: `numberOfHorizontalNeighboursWithSmokeCapacity()`.
Yeah, I know, its ugly long name. But better is ugly long that ugly short :P